### PR TITLE
New feature - run configurations

### PR DESCRIPTION
**Pull request by sillydan1**
This PR adds a new feature: Run configurations. Very much inspired by the jetbrains IDEs' feature of the same name. You can now run any program from inside the IDE.

**It would be much appreciated if I could get some help with the UI/UX design** specifically the run configurations-editor is not very pretty and could use some love

https://user-images.githubusercontent.com/4085393/200134260-b52d2600-0536-462d-b81e-c3b04ac8de4d.mov

## Features
 - Add run configurations: run any kind of program from inside H-UPPAAL

## Fixes / Minor improvements
 - Cleanup in classes that affect the run configurations feature
 - Buttons now have a subtle hover-effect, enhancing the user experience a bit
 - Updated `gson` to version `2.10` (notably, records can now be serialized/deserialized!)
 - add version information into the statusbar

## Missing Tasks
 - [x] Clicking "stop" whilst a run configuration is running should SIGTERM it, right now it just adds another thread
 - [x] Bump version number
 - [x] Run configurations should be able to provide additional environment variables to the command
 - [x] Test on MacOS
 - [x] Test on GNU/Linux
 - [ ] Test on Windows

For the moment run configurations will be stored in preferences. I personally think they should be saved in the project files instead. However, to keep the scope reasonable I have omitted this from this PR.

The output of a command will currently be put in the "Warnings" tab. This is not ideal, however I am also working on a complete rewrite of the CodeAnalysis/Logging system (I will make sure to keep it backwards compatible to ensure that the uppaal integration still works) so unless it's a really big issue (or easy to fix) I have also considered that "out-of-scope" for this PR.

### CODE REVIEW

File: src/main/java/dk/cs/aau/huppaal/controllers/HUPPAALController.java [hunk 1]
```diff
@@ -764,69 +731,50 @@ protected void interpolate(final double frac) {
...
 
     @FXML
     private void generateUppaalModelClicked() {
-        final Component mainComponent = HUPPAAL.getProject().getMainComponent();
-
+        var mainComponent = HUPPAAL.getProject().getMainComponent();
         if (mainComponent == null) {
-            HUPPAAL.showToast("Cannot generate UPPAAL file without a main component");
-            return; // We cannot generate a UPPAAL file without a main component
+            HUPPAAL.showToast("no main component");
```
Comment (Nielswps) [inline src/main/java/dk/cs/aau/huppaal/controllers/HUPPAALController.java]: Capitalize 'N' in "no"

File: src/main/java/dk/cs/aau/huppaal/controllers/RunConfigurationEditorController.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    private void initializeSavedConfigurationsList() {
+        var runConfigsJson = HUPPAAL.preferences.get(RunConfigurationPreferencesKeys.ConfigurationsList, "[]");
+        var currentConfig = HUPPAAL.preferences.get(RunConfigurationPreferencesKeys.CurrentlySelected, "");
+        var gson = new Gson();
+        List<RunConfiguration> runConfigurations = gson.fromJson(runConfigsJson, RunConfiguration.listTypeToken);
+        savedConfigurationsList.getItems().addAll(runConfigurations);
+        // TODO: set that c is selected if c.isPresent()
+        // var c = savedConfigurationsList.getItems().stream().filter(e -> e.name().equals(currentConfig)).findAny();
```
Comment (Nielswps) [inline src/main/java/dk/cs/aau/huppaal/controllers/RunConfigurationEditorController.java]: Is there a reason that this is not part of the PR? I think it would be nice to include it to avoid the configuration dropdown being "empty" when a configuration exists

File: src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java [hunk 3]
```diff
@@ -331,20 +318,154 @@ private void initializeColorSelector() {
...
 
+    private void initializeRunConfigExecuteButton() {
+        var colorIntensity = Color.Intensity.I800;
+        var color = Color.GREY_BLUE;
+        controller.runConfigurationExecuteButton.setRipplerFill(color.getTextColor(colorIntensity));
+
+        JFXTooltip tooltip;
+        var c = controller.runConfigurationPicker.getSelectionModel().getSelectedItem();
+        if(c == null || c.runConfiguration().isEmpty()) {
+            tooltip = new JFXTooltip("no run configuration is selected");
```
Comment (Nielswps) [inline src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java]: Capitalize 'N' in "no"

File: src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java [hunk 4]
```diff
@@ -331,20 +318,154 @@ private void initializeColorSelector() {
...
+            controller.runConfigurationExecuteButton.setEnabled(false);
+            controller.runConfigurationExecuteButton.setOpacity(0.3);
+            controller.runConfigurationExecuteButton.setOnMouseClicked(e -> HUPPAAL.showToast("no run configuration is selected"));
```
Comment (Nielswps) [inline src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java]: Capitalize 'N' in "no"

File: src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java [hunk 5]
```diff
@@ -331,20 +318,154 @@ private void initializeColorSelector() {
...
+                var dir = new File(config.executionDir);
+                if(!(dir.exists() && dir.isDirectory()))
+                    throw new Exception(String.format("'%s' does not exist or is not a directory", config.executionDir));
+                proc = rt.exec(config.program + " " + config.arguments,
+                                ArrayUtils.merge(sysEnv, config.environmentVariables.split(";")),
+                                dir);
+                var stdi = new BufferedReader(new InputStreamReader(proc.getInputStream()));
+                var stde = new BufferedReader(new InputStreamReader(proc.getErrorStream()));
+                controller.runConfigurationExecuteButtonIcon.setIconLiteral("gmi-stop");
+                controller.runConfigurationExecuteButtonIcon.setIconColor(javafx.scene.paint.Color.web("#ff7e79"));
```
Comment (Nielswps) [inline src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java]: Is there a reason not to use one of the colors specified in [colors.css](https://github.com/DEIS-Tools/H-Uppaal/blob/master/src/main/resources/colors.css) or one of the predefined colors in `javafx.scene.paint.Color`?

File: src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java [hunk 6]
```diff
@@ -331,20 +318,154 @@ private void initializeColorSelector() {
...
+    private void initializeRunConfigPicker() {
+        var runConfigsJson = HUPPAAL.preferences.get(RunConfigurationPreferencesKeys.ConfigurationsList, "[]");
+        var lastRunConfig = HUPPAAL.preferences.get(RunConfigurationPreferencesKeys.CurrentlySelected, "");
+        var runConfigurations = parseRunConfigurationsClearPreferencesIfFails(runConfigsJson);
+
+        // Add all the default things
+        controller.runConfigurationPicker.getItems().clear(); // Clear, because this function might be called again during runtime
+        controller.runConfigurationPicker.getItems().add(generateRunConfigurationEditButton());
+        if (runConfigurations.isEmpty())
+            controller.runConfigurationPicker.getItems().add(new RunConfigurationButton(Optional.empty(), new JFXButton("<no run configurations>")));
```
Comment (Nielswps) [inline src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java]: Can this string be shown as the default value for the dropdown?

File: src/main/java/dk/cs/aau/huppaal/presentations/RunConfigurationEditorPresentation.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+            return Optional.empty();
+        return Optional.of(c);
+    }
+
+    private void saveRunConfigurations() {
+        HUPPAAL.preferences.put(RunConfigurationPreferencesKeys.ConfigurationsList, gson.toJson(controller.savedConfigurationsList.getItems()));
+        HUPPAAL.showToast("Saved RunConfigurations");
+        if(onRunConfigsSaved != null)
+            onRunConfigsSaved.run();
+    }
```
Comment (Nielswps) [inline src/main/java/dk/cs/aau/huppaal/presentations/RunConfigurationEditorPresentation.java]: Consider setting the selected configuration to the newly saved one here

**Review by Nielswps — CHANGES_REQUESTED** [review]
What is the intended use case for this feature?
I personally would just use a terminal to run executables and wonder if this is outside the scope of the application. It is probably just because I do not have a lot of experience using the system, but an example of a use case would be nice

**Comment by sillydan1** [comment]
> What is the intended use case for this feature?
> I personally would just use a terminal to run executables and wonder if this is outside the scope of the application. It is probably just because I do not have a lot of experience using the system, but an example of a use case would be nice

The specific use-case would be that at our company we use a custom model compiler, debugger and a bunch of other utility tools (upload-scripts, unit-test runners etc.). Not all of these are implemented in java, so a direct java interface would be impractical (I would also argue that would be out of scope for the editor)

Model-designers are not necessarily developers themselves and they don't necessarily know how to use a terminal (especially on windows, this can be a problem) so this feature provides a generic way to integrate our tools, and makes it much easier to iterate on a model and get feedback from our tools. 

In combination with the [logging system refactor](https://github.com/DEIS-Tools/H-Uppaal/pull/52) tools can provide clickable links to exactly where an error has occurred - providing a seamless developer experience.

From the perspective of running a model-checker on your models, I would argue that it is more seamless to simply push a button in the editor and click on the problem, than to tab out, write the command, and interpret the standard-output manually to find the culprit component/syntax error/whatever went wrong. 

For this feature to be truly _complete_, run-configurations would have to be saved per-project and should be "shareable" (through git) rather than saved in the global preferences. However, I dont have much insight in the model-representation part of the codebase (right now, I'm slowly learning the architecture) so I haven't implemented that _yet_.

I can arrange a meeting about a use-case demonstration, if you'd like?

File: src/main/java/dk/cs/aau/huppaal/controllers/HUPPAALController.java (hunk 1 above)
Comment (sillydan1) [reply src/main/java/dk/cs/aau/huppaal/controllers/HUPPAALController.java]: Fixed in 4a4eaa25a8ec00e3b3fa4439a3e7078ea4d188fb

File: src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java (hunk 4 above)
Comment (sillydan1) [reply src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java]: Fixed in 4a4eaa25a8ec00e3b3fa4439a3e7078ea4d188fb

File: src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java (hunk 3 above)
Comment (sillydan1) [reply src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java]: Fixed in 4a4eaa25a8ec00e3b3fa4439a3e7078ea4d188fb

File: src/main/java/dk/cs/aau/huppaal/controllers/RunConfigurationEditorController.java (hunk 2 above)
Comment (sillydan1) [reply src/main/java/dk/cs/aau/huppaal/controllers/RunConfigurationEditorController.java]: I missed this during my cleanup step 😁 it has been fixed in a7913f33f47bd77265c2a9c6298149fc91a63d30 

Edit: 🤔 nevermind, it still doesn't select the thing.. I will look further into this

File: src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java (hunk 5 above)
Comment (sillydan1) [reply src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java]: My bad, I guess I forgot that they existed at the time 😄 

Fixed in a5ebb85ebe763dff085dccb674847ffba26269eb

File: src/main/java/dk/cs/aau/huppaal/presentations/RunConfigurationEditorPresentation.java (hunk 7 above)
Comment (sillydan1) [reply src/main/java/dk/cs/aau/huppaal/presentations/RunConfigurationEditorPresentation.java]: Fixed in 69b536fbcdcf24112773004796a5e97e20718c47

File: src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java (hunk 6 above)
Comment (sillydan1) [reply src/main/java/dk/cs/aau/huppaal/presentations/HUPPAALPresentation.java]: Fixed in 69b536fbcdcf24112773004796a5e97e20718c47

File: src/main/java/dk/cs/aau/huppaal/controllers/RunConfigurationEditorController.java (hunk 2 above)
Comment (sillydan1) [reply src/main/java/dk/cs/aau/huppaal/controllers/RunConfigurationEditorController.java]: Fixed 😁

**Review by Nielswps — APPROVED** [review]
Looks good :+1:  Thanks for clarifying the use case, I think it makes a lot of sense
